### ISSUE TITLE
using dataroom filter with reactive repositories

### ISSUE THREAD

**Issue by st-h** [crnk-project/crnk-framework#539]
I started to have a look at Crnk because I need json-api support for vertx. So far I have been able to figure out everything until I started to try out dataroom. Crnk logs warnings for all my registered repositories:

```
SecurityModule - no dataroom support for unknown repository xyz
```

I had a look at the [implementation](https://github.com/crnk-project/crnk-framework/blob/ff78f9d1a10b0d5e4f260f095bb1f5d6b8e8f4ca/crnk-security/src/main/java/io/crnk/security/SecurityModule.java#L201) and it seems that there currently is no support for reactive repositories yet since `ReactiveResourceRepository` does not inherit from `ResourceRepository`?

I'll be happy to take a stab at contributing a PR if we continue to use crnk, but most likely I will need a few pointers about what should be done as I am currently just starting  to look into Crnk.

**Comment by remmeier** [comment]
yes, true. For the basics most things are already in place. One just has to setup a reactive counter-part to the DataRoomResourceFilter. Potentially one may want to have a ReactiveDataRoomFilter as well to make the computation of the filter reactive.

**Comment by st-h** [comment]
@remmeier thanks for the reply. Does not sound like its very complicated to implement. However, so far I have been running into issues when integrating crnk with the vertx auth handlers, specifically together with vertx-web. The problems seems that the crnk `SecurityProvider` needs to be set up at a module level, however the `user` that vertx provides is only available at the `request` or `routingContext` level. Do you know if anyone has already made some plans regarding auth integration with vertx? It kind of looks to me that the concepts from Vertx-web and crnk will not work together nicely, but I might still be missing something fundamental. One could use vertx without the web extension, however that would also mean that a lot of what vertx-web provides will not be available (things like handling session data or cookies for instance)

**Comment by remmeier** [comment]
There are similar aspects with the spring, servlet, jaxrs, etc. integration. Certain things are only available from an outside context since we don't want to drag along all things in our repository (and other) contracts. For non-reactive systems this usually then ends up with thread- and request-scoped variables with the later being a automatically garbage-collected thread-scope as well. It is one of the benefit in the crnk architecture I think. Because client and servers are symmetric, can make use of the same repository contracts. context-related things are implement as independent, orthogonal concern.

Now crnk does have a concept of "context" which is provided through the `ResultFactory`. For reactive systems it is provided by `MonoResultFactory`. Most notably `ResultFactory.getContext` and `Resultfactory.attachContext(...)`. There are further thread-scoped counter-parts to bridge a bit to non-reactive systems. Internally it makes use of the `subscriberContext` to pass along the information together with the function calls. With that one get a hold onto the `HttpRequestContext` based on `HttpRequestContextProvider`, which brings it close to the requests and `vertx-web`, etc. So that is the path how to get hold to security-related information.

Hope I explained it more or less in an understandable way. Crnk is not 100% there yet, it is light and can integrate into basically everything, but not everything is as far along as for example the Spring integration. For example it might be benefitial for SecurityProvider to give a `Result<boolean isUserInRole(...)` rather than the current implementation. I guess you already have concrete ideas what you need. We could add those features to the current vertx example application.

**Comment by st-h** [comment]
@remmeier thanks for the reply. I don't think I fully understand everything right now, but that certainly are good pointers. I tried to follow the changes you described and it looks to me like we either will have to change the current api of the SecurityModule, or introduce a dedicated ReactiveSecurityModule (which should probably have a common interface together with the non reactive SecurityModule so that we at least can reuse some filters -  things like `RolePermissionRepository` will very likely need a reactive counterpart). Currently this starts to sound more like we need to design a reactive counterpart to the SecurityModule instead of just filling in some missing parts.

As for my requirements, I just need a rather simple JsonApi that supports authentication. We have a ember application, which has been server by vertx through a web socket. However, there were too many edge cases when trying to handle all data through the web socket (things like some browsers would not reopen the socket after the device had been in sleep mode). Ember on the other hand has very good support for JsonApi, so the missing piece here currently is JsonApi support in java/vertx, that's why I ended up here. Pretty much when a request comes in, we just need to look up the user by some token. With the user there pretty much is all the information needed to decide if the request should be denied or not. That should probably not be too complicated to get going.

I can put aside a day or two to add things that vertx support is lacking in crnk, however by now I doubt that this will be sufficient to make everything available that other non reactive implementation currently support. But I think it might be a good idea to agree on some architecture details, before adding code.

**Comment by remmeier** [comment]
Just the SecurityModule is fine. I think 90% of it will share the code base. I can look into that area as well.

For the SecurityProvider and ResourceFilter should gain a context attribute (like SecurityProviderContext and ResourceFilterContext), which in turn gives access to e.g. `QueryContext`. And vertx can make its things then available in there. It is a breaking change, but not unexpected and would fit in the current 3.0 release that is happening anyway.

I don't think that the various interfaces in the area should become "reactive". They are performance sensitive, so the information from the passed context must be sufficient to come to an immedidate decision. If something needs to be resolved somewhere that takes time, it has to happen at an earlier stage when populating the context.

**Comment by st-h** [comment]
@remmeier sorry, I'll need some clarification

> SecurityProvider to give a Result<boolean isUserInRole(...)

If I make that change, I will also have to do that in the SecurityModule itself (as it contains an implementation of that interface). So that will either be a change that effects non reactive implementations of crnk, or we will have to add dedicated reactive things like a reactive SecurityModule and a reactive SecurityProvider. Would you prefer the first or the latter?

**Comment by remmeier** [comment]
I would really change those two interfaces. It will then also affect non-reactive users. But since it is in the engine part of crnk, it will be a minority with an easy fix. Usually it is a no go. But as part of 3.0 it would be good to have that finally sorted out.

If you have time, would be great. You could also start by just modifying the new interfaces and then we have a look together. I'm not 100% sure how much it affects the remaining engine part. But I can help out in the area. ResourceFilterDirectory already works with QueryContext, so I think it is mostly ok already, just bringing QueryContext one step further.

**Comment by st-h** [comment]
@remmeier awesome. I opened a [PR](https://github.com/crnk-project/crnk-framework/pull/541) and as first step changed the SecurityProvider::isUserInRole return type. Can you have a look, if this is what you had in mind?

 However I am having trouble compiling the project because of errors in creating the docs
```
Exception in thread "main" org.asciidoctor.gradle.backported.AsciidoctorRemoteExecutionException: Error running Asciidoctor whilst attempting to process crnk-framework/crnk-documentation/src/docs/asciidoc/cheatsheet.adoc using backend pdf
...
Caused by: org.jruby.exceptions.RaiseException: (LoadError) load error: jopenssl/load -- java.lang.StringIndexOutOfBoundsException: String index out of range: 0
```
 as well as missing xml annotations (which might be related to the my jvm)
```
package javax.xml.bind.annotation does not exist
import javax.xml.bind.annotation.XmlRootElement;
```
I will need to look into those, but if the build should fail for now, that's why.

**Comment by remmeier** [comment]
we will make use of Java 8 for compilation. That javax.xml.bind has been removed later I think.

**Comment by remmeier** [comment]
got my things done and just merged, will proceed with this one
